### PR TITLE
Hotfix 1484 chrome exception signup

### PR DESCRIPTION
**Pull request by shane-tomlinson**
Swallow exceptions that Chrome throws when writing to localStorage if cookies are disabled.

I am not fully happy with this solution as it is a bandaid instead of a true fix for the root cause. I am writing a fuller solution against the current dev to show the cookies disabled error screen as soon as the user lands on the majority of main site pages.

issue #1484.

### CODE REVIEW

**Comment by benadida** [comment]
should you maybe console.log something so it's not failing silently?

**Comment by shane-tomlinson** [comment]
@benadida - good idea.  I will add console logging.

File: resources/static/shared/storage.js [hunk 1]
```diff
@@ -175,14 +192,25 @@ BrowserID.Storage = (function() {
...
 
   function generic2KeyGet(namespace, key) {
-    var allInfo = JSON.parse(storage[namespace] || "{}");
-    return allInfo[key];
+    try {
+      var allInfo = JSON.parse(storage[namespace] || "{}");
+      return allInfo[key];
+    } catch(e) {
+      logError("error reading from storage: " + e);
+      return undefined;
```
Comment (ozten) [inline resources/static/shared/storage.js:212]: Cool, the semantics of returning undefined look the same, assuming the old code it was possible to ask for a key that didn't exist in allInfo.

File: resources/static/shared/storage.js [hunk 2]
```diff
@@ -28,6 +28,17 @@ BrowserID.Storage = (function() {
     }
   }
 
+  function logError(msg) {
+    try {
+      console.log(msg);
+    } catch(e) {
+      // Do nothing. The console.log is inside of a try/catch because some
```
Comment (ozten) [inline resources/static/shared/storage.js:35]: This is fine. I'm surprised we didn't have something like this in the code already:

```
if (window.console && console.log ) console.log(msg);
```

File: resources/static/shared/storage.js [hunk 3]
```diff
@@ -175,14 +192,25 @@ BrowserID.Storage = (function() {
...
-    allInfo[key] = value;
-    storage[namespace] = JSON.stringify(allInfo);
+    try {
+      var allInfo = JSON.parse(storage[namespace] || "{}");
+      allInfo[key] = value;
+      storage[namespace] = JSON.stringify(allInfo);
+    } catch(e) {
+      // Do nothing, Chrome, Safari and FF will blow up here if cookies are
+      // disabled.
+      logError("error writing to storage: " + e);
```
Comment (ozten) [inline resources/static/shared/storage.js:202]: Can the app continue without this storage? If not, we should have a global flag somewhere to bail and show an error message to the user.

If this is just a performance enhancement or we have other workarounds, than this looks good.

**Comment by ozten** [comment]
This is looking really good. One question before I'm ready to r+ or r-.

File: resources/static/shared/storage.js (hunk 1 above)
Comment (shane-tomlinson) [inline resources/static/shared/storage.js:212]: I could remove that return undefined, eh?

File: resources/static/shared/storage.js (hunk 2 above)
Comment (shane-tomlinson) [inline resources/static/shared/storage.js:35]: in PR-1489(https://github.com/mozilla/browserid/pull/1489), I am adding a helpers.log so we can have this sort of logic in a single location - https://github.com/mozilla/browserid/pull/1489/files#L1R70

File: resources/static/shared/storage.js (hunk 3 above)
Comment (shane-tomlinson) [inline resources/static/shared/storage.js:202]: @ozten - the app cannot continue without storage.  I agree with your concern here and created PR-1489 to globally address this for all pages on the main site (minus the splash page) - if cookies are disabled (making localStorage unavailable), then an error is displayed to the user immediately upon landing on the page.

**Comment by shane-tomlinson** [comment]
@ozten, @benadida - The aim of this pull request is to prevent Chrome from excepting on the signin page whenever cookies are disabled.  What we do in the signin, signup, and forgot password pages is every time the user updates the email address, we write the email address to localStorage.  If cookies are disabled in Chrome, an exception is thrown.  With PR-1489 (https://github.com/mozilla/browserid/pull/1489), I avoid this situation entirely.  An error message is displayed to the user as soon as they land on the page if their cookies are disabled.

Since this was a hotfix that was aimed at fixing the current beta and since it has been decided to let this exception continue to occur, I feel a more complete solution should be used in place of this one.

Perhaps _all_ reads and writes to localStorage should occur through a set of accessors that contain try/catch blocks.  In this way, we could write console messages from a single place instead of sprinkling them throughout storage.js

**Comment by ozten** [comment]
This sounds like a good cleanup idea! hotfix vs long term solution makes sense.

**Pull request closed without merging**